shared/src/commonMain/kotlin/co/nimblehq/ic/kmm/suv/data/remote/apiclient/core/ApiClient.kt

```diff
@@ -0,0 +1,110 @@
+import io.ktor.client.statement.*
+import io.ktor.http.*
+import io.ktor.serialization.kotlinx.json.*
+import kotlinx.coroutines.flow.*
+import kotlinx.serialization.json.Json
+
+class ApiClient(engine: HttpClientEngine, tokenLocalDataSource: TokenLocalDataSource? = null) {
+
+    val httpClient: HttpClient
+    val json = Json {
+        prettyPrint = true
+        isLenient = true
+        encodeDefaults = true
+        ignoreUnknownKeys = true
+    }
+
+    private val tokenLocalDataSource: TokenLocalDataSource?
+
+    init {
+        this.tokenLocalDataSource = tokenLocalDataSource
+        Napier.takeLogarithm()
+        Napier.base(DebugAntilog())
+        httpClient = HttpClient(engine = engine) {
+            install(Logging) {
+                level = ALL
+                logger = object : Logger {
+                    override fun log(message: String) {
+                        Napier.log(LogLevel.DEBUG, message = message)
+                    }
+                }
+            }
+
+            install(ContentNegotiation) {
+                json(json)
+            }
+
+            tokenLocalDataSource?.let {
+                install(Auth) {
+                    bearer {
+                        loadTokens {
+                            val token = it.getToken()
+                            token?.let {
+                                BearerTokens(it.accessToken, it.refreshToken)
+                            }
+                        }
+
+                        sendWithoutRequest { request ->
+                            request.url.host == Url(BuildKonfig.BASE_URL).host
+                        }
+
+                        refreshTokens {
+                            val token = refreshToken(oldTokens?.refreshToken ?: "").last()
+                            tokenLocalDataSource.save(token)
+                            tokenLocalDataSource.getToken()?.let {
+                                BearerTokens(it.accessToken, it.refreshToken)
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    inline fun <reified T> body(builder: HttpRequestBuilder): Flow<T> {
+        return flow {
+            val body = httpClient.request(
+                builder.apply {
+                    contentType(ContentType.Application.Json)
+                }
+            ).bodyAsText()
+            try {
+                val data = JsonApi(json).decodeFromJsonApiString<T>(body)
+                emit(data)
+            } catch (e: JsonApiException) {
+                val message = e.errors.first().detail
+                throw AppError(message)
+            }
+        }
+    }
+
+    private fun refreshToken(token: String): Flow<TokenApiModel> {
+        val body = RefreshTokenApiBody(refreshToken = token)
+        return body(HttpRequestBuilder().apply {
+            path("/v1/oauth/token")
+            method = HttpMethod.Post
+            setBody(body)
+        })
+    }
```

Comment:
I think for a better network layer design, we should move this refresh token API to be listed in `TokenRemoteDataSourceImpl`. It still uses `apiClient` to build the API request so there would be no issue 💭 

Then the idea is to execute `refresh token` request via a use case (RefreshTokenUseCase) (more preferred) or repository 💭

Reply:
Fixed in https://github.com/suho/kmm-ic/pull/94/commits/fca5d8acb0b263ec50661cd79bd5560e4cad0f8f

I will use `TokenRemoteDataSource`, as we already have `TokenLocalDataSource` in this clase